src/common/procedure/src/local/runner.rs

```diff
@@ -66,5 +117,527 @@ impl Runner {
+        match self.procedure.execute(ctx).await {
+            Ok(status) => {
+                if status.need_persist() && self.persist_procedure().await.is_err() {
+                    return ExecResult::RetryLater;
+                }
+
+                match status {
+                    Status::Executing { .. } => (),
+                    Status::Suspended { subprocedures, .. } => {
+                        self.on_suspended(subprocedures).await;
+                    }
+                    Status::Done => {
+                        if self.commit_procedure().await.is_err() {
+                            return ExecResult::RetryLater;
+                        }
+
+                        self.done();
+                        return ExecResult::Done;
+                    }
+                }
+
+                ExecResult::Continue
+            }
+            Err(e) => {
+                logging::error!(
+                    e;
+                    "Failed to execute procedure {}-{}",
+                    self.procedure.type_name(),
+                    self.meta.id
+                );
+
+                self.meta.set_state(ProcedureState::Failed);
+
+                // Write rollback key so we can skip this procedure while recovering procedures.
+                if self.rollback_procedure().await.is_err() {
+                    return ExecResult::RetryLater;
+                }
+
+                ExecResult::Failed(e)
+            }
+        }
+    }
+
+    /// Submit a subprocedure with specific `procedure_id`.
+    fn submit_subprocedure(&self, procedure_id: ProcedureId, mut procedure: BoxedProcedure) {
+        let mut step = 0;
+        if let Some(loaded_procedure) = self.manager_ctx.load_one_procedure(procedure_id) {
+            // Try to load procedure state from the message to avoid re-run the subprocedure
+            // from initial state.
+            assert_eq!(self.meta.id, loaded_procedure.parent_id.unwrap());
+
+            // Use the dumped procedure from the procedure store.
+            procedure = loaded_procedure.procedure;
+            // Update step number.
+            step = loaded_procedure.step;
+        }
+
+        if self.manager_ctx.contains_procedure(procedure_id) {
+            // If the parent has already submitted this procedure, don't submit it again.
+            return;
+        }
+
+        let meta = Arc::new(ProcedureMeta {
+            id: procedure_id,
+            lock_notify: Notify::new(),
+            parent_id: Some(self.meta.id),
+            child_notify: Notify::new(),
+            lock_key: procedure.lock_key(),
+            exec_meta: Mutex::new(ExecMeta::default()),
+        });
+        let runner = Runner {
+            meta: meta.clone(),
+            procedure,
+            manager_ctx: self.manager_ctx.clone(),
+            step,
+            store: self.store.clone(),
+        };
+
+        self.manager_ctx.insert_procedure(meta);
```

Comment:
Is it possible that when you `insert_procedure` here, `manager_ctx` already contains the same procedure?

Reply:
Good point! Normally this won't be a problem as ProcedureId is an UUID. But there still has chance that different procedures use the same ProcedureId to submit a subprocedure. I'd add a check to `insert_procedure`.